software/glasgow/target/toolchain.py

```diff
@@ -0,0 +1,321 @@
+    @abstractmethod
+    def identifier(self):
+        """Unique tool identifier.
+
+        Returns an array of 16 bytes that uniquely identifies the behavior of this particular tool
+        in its entirety, but has no other meaning. Typically implemented by hashing the binary and
+        its data files.
+        """
+
+    def __repr__(self):
+        return f"<{self.__class__.__module__}.{self.__class__.__name__} {self.name}>"
+
+
+
+class WasmTool(Tool):
+    PREFIX = "yowasp-"
+
+    @property
+    def python_package(self):
+        if self.name == "yosys" or self.name.startswith("nextpnr-"):
+            return self.PREFIX + self.name
+        if self.name == "icepack":
+            return self.PREFIX + "nextpnr-ice40"
+        if self.name == "ecppack":
+            return self.PREFIX + "nextpnr-ecp5"
+        raise NotImplementedError(f"Python package for tool {self.name} is not known")
+
+    @property
+    def available(self):
+        try:
+            importlib.metadata.metadata(self.python_package)
+            return True
+        except importlib.metadata.PackageNotFoundError:
+            return False
+
+    @property
+    def command(self):
+        if self.available:
+            basename = self.PREFIX + self.name
+            # We cannot assume that the command is on PATH and accessible by its basename. This
+            # will not be true when Glasgow is running from a pipx virtual environment (which isn't
+            # activated when the `glasgow` script is run). Also, our build environment does not
+            # even *have* PATH.
+            return os.path.join(sysconfig.get_path('scripts'), basename)
+
+    @property
+    def version(self):
+        if self.available:
+            # Running Wasm tools for the first time can incur a significant delay, so use
+            # the version from the Python package metadata (which is guaranteed to be the same).
+            # This makes querying the version at least as fast as for the native tools.
+            return (*importlib.metadata.version(self.python_package).split("."),)
+
+    @property
+    def identifier(self):
+        if self.available:
+            hasher = hashlib.blake2s()
+            for file_entry in importlib.metadata.files(self.python_package):
+                if file_entry.hash is None:
+                    continue # RECORD, *.pyc, etc
+                hasher.update(file_entry.hash.value.encode("utf-8"))
+            return hasher.digest()[:16]
+
+
+class NativeTool(Tool):
+    @staticmethod
+    def get_output(args):
+        return subprocess.run(args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            encoding="utf-8").stdout.strip()
+
+    @property
+    def available(self):
+        return self.command is not None
+
+    @property
+    def command(self):
+        return shutil.which(os.environ.get(self.env_var_name, self.name))
+
+    @property
+    def version(self):
+        if self.available:
```

Comment:
You changed from early return... is this a preference for you? (elsewhere too... I'd prefer early return, but won't block)

```suggestion
        if not self.available:
            return
```

Reply:
I normally prefer early return but here in particular it made the code way longer vertically than it needed to be and somewhat harder to read.